clients/src/main/java/org/apache/kafka/common/network/ConnectionRegistry.java

```diff
@@ -0,0 +1,258 @@
+            String clientSoftwareVersion,
+            ListenerName listenerName,
+            SecurityProtocol securityProtocol,
+            InetAddress clientAddress,
+            KafkaPrincipal principal) {
+
+        ConnectionMetadata connection = new ConnectionMetadata(
+            clientId,
+            clientSoftwareName,
+            clientSoftwareVersion,
+            listenerName,
+            securityProtocol,
+            clientAddress,
+            principal
+        );
+
+        connections.put(connectionId, connection);
+
+        incCounterFor(connection.clientSoftwareName(), connection.clientSoftwareVersion());
+
+        return connection;
+    }
+
+    /**
+     * Updates the client software name and version.
+     *
+     * Note that the method is synchronized to guard the counters.
+     */
+    public synchronized ConnectionMetadata updateClientSoftwareNameAndVersion(
+            String connectionId,
+            String clientSoftwareName,
+            String clientSoftwareVersion) {
+
+        ConnectionMetadata connection = connections.get(connectionId);
+
+        if (connection != null) {
+            decCounterFor(connection.clientSoftwareName(), connection.clientSoftwareVersion());
+            connection.updateClientSoftwareNameAndVersion(clientSoftwareName, clientSoftwareVersion);
+            incCounterFor(connection.clientSoftwareName(), connection.clientSoftwareVersion());
+        }
+
+        return connection;
+    }
+
+    /**
+     * Removes metadata about a connection.
+     *
+     * Note that the method is synchronized to guard the counters.
+     */
+    public synchronized void remove(String connectionId) {
+        ConnectionMetadata connection = connections.remove(connectionId);
+
+        if (connection != null) {
+            decCounterFor(connection.clientSoftwareName(), connection.clientSoftwareVersion());
+        }
+    }
+
+    /**
+     * Closes the registry.
+     */
+    public synchronized void close() {
+        for (CounterMetric counter : counters.values()) {
+            metrics.removeMetric(counter.name());
+        }
+        metrics.removeMetric(connectedClientsMetricName);
+        metrics.removeMetric(connectionsMetricName);
+        counters.clear();
+        connections.clear();
+    }
+
+    /**
+     * Gets metadata about a connection.
+     *
+     * Note that the method is NOT synchronized. A lookup is made for each request in the
+     * processor to enrich the request with the metadata (e.g. client software name and version).
+     * We have made the choice to not synchronize this to minimize the contention between the
+     * processor threads, risking that metadata could be stale or inconsistent for few requests. This
+     * should rarely happen in practice because the metadata is only updated by the ApiVersionsRequest
+     * and there are normally no concurrent requests within the connection.
+     */
+    public ConnectionMetadata get(String connectionId) {
```

Comment:
This is wrong.  Unsynchronized access to a map could cause more than just "stale or inconsistent" data.  It could cause null pointer exceptions or other issues.  We cannot access this without synchronization.